### ISSUE TITLE
LazyLinear troubles

### ISSUE THREAD

**Issue by rfechner** [patrick-kidger/equinox#949]
Hi :) I'm new to equinox and tried hard to get an implementation of a lazily initialized linear layer going, however without any success. I coudn't really find any resource on lazy initialization in equinox, hence I've tried to hack it this:

```python
class LazyLinear(eqx.Module):
    """
        Lazy Linear layer which is only fully defined after the first call to the
        __call__ function. Infers the `in_features` keyword-argument necessary to define
        a linear layer. 

        Used to make convolutional architectures flexible, s.t. different input sizes, kernel sizes, paddings, strides etc may be tested.
    """
    key : jr.PRNGKey
    initialized : bool
    linear : eqx.nn.Linear
    out_features : int
    use_bias : bool

    def __init__(self, *, key, out_features, use_bias=True, **kwargs):
        super().__init__(**kwargs)

        self.key = key
        self.initialized = False

        # placeholder linear layer
        self.linear = eqx.nn.Linear(1,1, key=key)
        self.out_features = out_features
        self.use_bias = use_bias

    def __call__(self, x, *, key=None, **kwargs):

        if not self.initialized:
            in_features = x.shape[-1]
            linear = eqx.nn.Linear(in_features=in_features, out_features=self.out_features, use_bias=self.use_bias, key=self.key)
            

            self.__dict__['initialized'] = True
            self.__dict__['linear'] = linear

        return self.linear(x)
```

Python allows me to mutate the frozen dataclass by using these hacks, however they seem to not be carried over once i leave the scope (the id of the nested linear layer changes every time i use the call function, i suppose its re-created every single time i use `__call__` . My best guess at the moment is that as eqx.Module is a frozen dataclass, there is some optimization happening which prohibits the local changes from being applied.

Alternatively, I've come up with the solution of having an `infer_in_features` function, which takes the previous layers of a network, (which also have to be a eqx.Module) then running a dummy data through to infer the shape. This works, but is ugly.

The question now is: Is there a correct way of handling this issue? 

PS. What would be the correct way of handling (de-) serialization with this LazyLinear approach? I've tried something like:

```python
custom_isleaf = lambda x: isinstance(x, LazyLinear)

def custom_serialize(f, x):
    if isinstance(x, LazyLinear):
        pickle.dump(obj=x, file=f)
    else:
        eqx.default_serialise_filter_spec(f, x)

def custom_deserialize(f, x):
    if isinstance(x, LazyLinear):
        return pickle.load(f)
    else:
        return eqx.default_deserialise_filter_spec(f, x)
```

and then pass them as kwargs in:

```python
eqx.tree_serialise_leaves(path_or_file=file, pytree=pytree, filter_spec=custom_serialize, is_leaf=custom_isleaf)
```

Thanks for any comments, pointers or corrections. Also thanks for developing equinox, i like it :) -Richard

**Comment by johannahaffner** [comment]
Hi Richard, 

from my knowledge of Equinox and JAX, this really runs counter to the design of both. JAX expects constant array sizes at compile time and re-compiles if these change, and equinox modules are immutable by design - personally I find this a reliable bug prevention method :) 

Since you want to test different shapes (features, layers, ...), I assume you are fine with re-compiling for each of these. To make this easy, I suggest you write yourself a helper function 

```python

class Linear(eqx.Module):
      ....

def make_linear(hyperparameters):
      return Linear(**hyperparameters)
```

and then you create your scripts such that each of these model variants gets its own jitted region. 

Alternatively, if you do not want to re-compile, I suggest you look into any of the following: 
- masked arrays, with a mask that is specified at runtime
- custom filter_specs, which allow you to [freeze some parameters](https://docs.kidger.site/equinox/examples/frozen_layer/), these could allow you to train layers selectively
- [tree surgery](https://docs.kidger.site/equinox/api/manipulation/#equinox.tree_at), to replace layers with dummy versions (the effect should be similar to masking these out, but the approach is different).

 Depending on the complexity of your setup, you may need more than one of the above.

**Comment by rfechner** [comment]
Hi Johanna, thanks for the quick reply :)

Some context: Im using the lazy linear layer in something like:

```python
class ResNet1D(BaseModel):
    features : eqx.Module
    lin1: LazyLinear
    classify : eqx.nn.Linear

    def __init__(self, *, key, in_channels: int, num_classes: int, in_feature_shape : List[int], num_blocks: int, hidden_dim: int, normalize: bool = False):
        """
            NOTE: `in_feature_shape` remains a positional argument, which should be defined and carried until we've asserted
            that `LazyLinear` is working as intended.
        """
        
        feature_key, lin1_key, out_key = jax.random.split(key, 3)
        
        self.features = ResNetBackbone(
            in_channels=in_channels,
            num_blocks=num_blocks,
            hidden_dim=hidden_dim,
            normalize=normalize,
            feature_key=feature_key
        )

        self.lin1 = LazyLinear(out_features=hidden_dim, key=lin1_key)
        self.classify = eqx.nn.Linear(in_features=hidden_dim, out_features=num_classes, key=out_key)

    def __call__(self, x, state, *, key=None):
        x, state = self.features(x, state, key=key)

        x = x.reshape(-1)
        x = self.lin1(x)
        x = jax.nn.relu(x)
        x = self.classify(x)
        x = jax.nn.log_softmax(x)
        
        return x, state
```
I'm using a configuration manager to set all other hyperparameters. However, when choosing strides, depth etc as variable inputs to my class, calculating the input shape to the first linear layer after the backbone becomes tedious (im lazy). Not only this, but I'd have to specify those parameters for every dataset (they have variable feature and spatial dimensions) which makes using something like a lazily initialized linear layer way more attractive. I'm basically just looking for a mechanism which allows me to infer the "in_features" of the first linear layer.

What do you mean with "at compile time"? Is this before my first

 ```python
jax.vmap(model)(x)
``` 
call? Doesn't jax pass some tracers through my model before i actually get to call it on my data? Anyways if i need to specify the shapes before the `__call__`  is even used, then I can probably forget about any lazy initializations which don't involve explicitly calling something like an initialize function and pre-compiling a sub-module of my larger module pre-pended to the lazylinear layer...

Thanks again for the ideas and answering :) - Richard

**Comment by rfechner** [comment]
To illustrate my current workaround: I'm calling this function to map the "backbone" field of my module and the input shape to the input shape of my first linear layer after the backbone:

```python
def infer_shape(module_class : Callable, module_kwargs : Dict[str, Any], input_shape : List[int], flatten=True) -> Tuple[int, ...]:
    """
        Given a list of modules, we infer the shape of the output.
    """

    def maybe_prepend_feature_dimension(inp : List[int]):
        """
            for equinox's CNNs we need an explicit feature dimension apart from the
            spatial dimension. Our data should have the shape [features, spatial]
        """
        shape = deepcopy(inp)
        if len(inp) == 1:
            # pre-pend equinox feature dimension.
            shape = [1] + shape
        return shape
    
    shape = maybe_prepend_feature_dimension(input_shape)

    # preparing the tracing
    key, tracer_key = jr.split(jr.PRNGKey(0))
    tracer = jnp.expand_dims(jr.normal(tracer_key, shape=shape), axis=0) # shape : [batch, feature, spatial]
    
    # map the model onto the dummy
    partial_model, state = eqx.nn.make_with_state(module_class)(**module_kwargs)
    batched_model = jax.vmap(partial(partial_model, key=key), in_axes=(0, None), out_axes=(0, None), axis_name='batch')
    
    batched_output, state = batched_model(tracer, state)
    output = batched_output.squeeze(0) # remove batch dim

    if flatten:
        # reduces (a, b, c) to (a * b * c,)
        return (reduce(lambda a, b: a * b, output.shape, 1),)

    return output.shape
```

**Comment by johannahaffner** [comment]
Hi Richard, and sure!

> What do you mean with "at compile time"? Is this before my first
> 
> jax.vmap(model)(x)
> call? Doesn't jax pass some tracers through my model before i actually get to call it on my data? 

JAX traces the model _for inputs of the shape you pass to it at compile time_, which is usually the first time the function is called. When the input shapes change, JAX will re-trace and re-compile. The computation graph it creates is not shape-agnostic - you can look at a [simple jaxpr](https://docs.jax.dev/en/latest/jaxpr.html#handling-pytrees) example to see that it really parses things down to the level of: "take element i from x and add it to element j from y". 

Compilation is much more expensive than running a compiled model, so you typically want to avoid recompiling.

> Anyways if i need to specify the shapes before the `__call__` is even used, then I can probably forget about any lazy initializations which don't involve explicitly calling something like an initialize function and pre-compiling a sub-module of my larger module pre-pended to the lazylinear layer...

...which brings me to: yes, at least in the way you were doing it for now. Some clever padding trick might be possible, but you'll have to be quite careful in how you set it up.

**Comment by rfechner** [comment]
Alright. Before closing this issue, I'd like to ask if this sort of feature is planned in a future version of equinox?

**Comment by rfechner** [comment]
Right, thanks for the hint ^^ Seems like exactly the kind of function i was looking for. For completeness sake, here is an updated version of the infer_shape function which seems to work for my (simple) needs:

```python
def infer_shape(model : eqx.Module, model_state, input_shape : List[int], flatten=True) \
    -> Tuple[int, ...]:
    """
        Given a an eqx.Module, we infer the shape of its output.

        NOTE: Yes, i tried lazy initializing a eqx.nn.Linear layer first. For discussion, see:
        https://github.com/patrick-kidger/equinox/issues/949
    """

    def maybe_prepend_feature_dimension(inp : List[int]):
        """
            for equinox's CNNs we need an explicit feature dimension apart from the
            spatial dimension. Our data should have the shape [features, spatial]
        """
        shape = deepcopy(inp)
        if len(inp) == 1:
            # pre-pend equinox feature dimension.
            shape = [1] + shape
        return shape
    
    shape = maybe_prepend_feature_dimension(input_shape)

    # preparing the tracing
    key, tracer_key = jr.split(jr.PRNGKey(0))
    tracer = jnp.expand_dims(jr.normal(tracer_key, shape=shape), axis=0)
    
    # eqx.nn.BatchNorm only works when axis_name is defined in vmap
    batched_model = jax.vmap(partial(model, key=key), in_axes=(0, None), out_axes=(0, None), axis_name='batch')
    
    out, model_state = jax.eval_shape(batched_model, tracer, model_state)
    out_shape = out.shape[1:] # remove batch dimension

    if flatten:
        # reduces (a, b, c) to (a * b * c,)
        return (reduce(lambda a, b: a * b, out_shape, 1),)
    
    return out_shape
```

and code to show usage inside my custom module:

```python
class ResNet1D(BaseModel):
    features : eqx.Module
    lin1: eqx.nn.Linear
    classify : eqx.nn.Linear

    def __init__(self, *, key, in_channels: int, num_classes: int, in_feature_shape : List[int], num_blocks: int, hidden_dim: int, normalize: bool = False):
        
        feature_key, lin1_key, out_key = jax.random.split(key, 3)
        
        self.features, tmp_state = eqx.nn.make_with_state(ResNetBackbone)(
            in_channels=in_channels,
            num_blocks=num_blocks,
            hidden_dim=hidden_dim,
            normalize=normalize,
            feature_key=feature_key
        )

        input_shape : Tuple[int, ...] = \
            infer_shape(self.features, tmp_state, input_shape=in_feature_shape)
        
        self.lin1 = eqx.nn.Linear(in_features=input_shape[0], out_features=hidden_dim, key=lin1_key)
        ...
```
